### PR TITLE
feat(coderd): notify when workspace is marked as dormant

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Related to https://github.com/coder/team-coconut/issues/7

### CODE REVIEW

File: coderd/workspaces.go
Comment (dannykopping) [inline coderd/workspaces.go:973]: Is this API only ever called by user-initiated actions? I had a brief look and it seems to be, but maybe you have more context. If that's the case, we don't need this constant value - we can just always say "requested by user" in the template.

File: coderd/dormancy/notifications.go
Comment (dannykopping) [inline coderd/dormancy/notifications.go:1]: Since dormancy is an enterprise-only feature, you should consider moving this to the existing `dormancy` package in `enterprise/coderd/dormancy`.

File: coderd/workspaces_test.go
Comment (dannykopping) [inline coderd/workspaces_test.go:3535]:
```suggestion
			t.Cleanup(cancel)
```

File: coderd/workspaces_test.go
Comment (dannykopping) [inline coderd/workspaces_test.go:3638]: Rather than copying this from `coderd/provisionerdserver/provisionerdserver_test.go`, let's move this into the `notifications` package under test utilities.

File: coderd/workspaces.go
Comment (dannykopping) [inline coderd/workspaces.go:966]: Is this the only place this will be called? If so, I don't see the need for the helper method or the struct.

File: coderd/database/migrations/000221_notifications.up.sql
Comment (dannykopping) [inline coderd/database/migrations/000221_notifications.up.sql:72]: You should move this out into its own migration; customers who have already run this migration will not get the new template.

File: coderd/workspaces_test.go
Comment (dannykopping) [inline coderd/workspaces_test.go:3640]: 👍 nice

File: coderd/workspaces_test.go
Comment (dannykopping) [inline coderd/workspaces_test.go:3510]: Nice tests 👍 
Nit: you could add some more comments to separate `GIVEN`/`WHEN`/`THEN` and explain what you're doing to make them more readable, but they're quite simple and clean already.

**Review by dannykopping — COMMENTED** [review]
Good work, left a few thoughts

File: coderd/workspaces.go
Comment (BrunoQuaresma) [reply coderd/workspaces.go:973]: I need to implement this notification in two places: the lifecycle executor and the scheduler.

File: coderd/workspaces.go
Comment (BrunoQuaresma) [reply coderd/workspaces.go:966]: Nops, we are going to call them in the lifecycle executor and the scheduler.

File: coderd/database/migrations/000223_dormancy_notification_template.up.sql
Comment (dannykopping) [inline coderd/database/migrations/000223_dormancy_notification_template.up.sql]: Are you up-to-date with `main`? We're already up to `225`: https://github.com/coder/coder/tree/main/coderd/database/migrations

File: coderd/workspaces.go
Comment (dannykopping) [inline coderd/workspaces.go:960]: We shouldn't return in this scenario

File: coderd/autobuild/lifecycle_executor.go
Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor.go:223]: I think this should go after the error is checked to be `nil`.

File: enterprise/coderd/schedule/template.go
Comment (dannykopping) [inline enterprise/coderd/schedule/template.go:169]: We'll have to do this outside of the transaction. If the transaction fails and needs to be rolled back, then the notification will still be sent and will indicate an action which didn't complete.

File: enterprise/coderd/schedule/template_test.go
Comment (dannykopping) [inline enterprise/coderd/schedule/template_test.go:301]: Please pass a real logger here in case something fails in the tests.

```go
slogtest.Make(t, &slogtest.Options{IgnoreErrors: true}).Leveled(slog.LevelDebug)
```

File: coderd/database/dbauthz/dbauthz.go
Comment (dannykopping) [inline coderd/database/dbauthz/dbauthz.go:3530]: Be careful that the semantics didn't change here since you're moving away from `fetchAndExec`.

File: coderd/workspaces.go
Comment (BrunoQuaresma) [reply coderd/workspaces.go:960]: 💯 🤦

File: coderd/database/dbauthz/dbauthz.go
Comment (BrunoQuaresma) [reply coderd/database/dbauthz/dbauthz.go:3530]: What do you mean? 🤔

Comment (dannykopping) [reply coderd/database/dbauthz/dbauthz.go:3530]: Previously the function was using `fetchAndExec` but now you've replaced that, so just take care that this didn't change the behaviour of this function.

Comment (BrunoQuaresma) [reply coderd/database/dbauthz/dbauthz.go:3535]: Now it is returning the affected workspaces, which I think is okay and is related to this method signature. Wdyt?

File: enterprise/coderd/dormancy/notifications.go
Comment (BrunoQuaresma) [reply enterprise/coderd/dormancy/notifications.go:41]: That is completely true. I don't know why I didn't see that before.

File: coderd/database/migrations/000226_dormancy_notification_template.up.sql
Comment (dannykopping) [inline coderd/database/migrations/000226_dormancy_notification_template.up.sql:3]: I think we'll need to add some detail here about what "dormant" means; perhaps a link to the docs and how to fix it.

File: coderd/schedule/mock.go
Comment (dannykopping) [inline coderd/schedule/mock.go:16]: Let's remove these loggers since they're not needed anymore.

**Comment by dannykopping** [comment]
@BrunoQuaresma I tried to move this along while you were out, but I didn't have much time.
I added https://github.com/coder/coder/pull/13868/commits/467a797edcd927d66c47f0ef44b52f0bee22f05e which simplifies the implementation, although I have not had time to test it thoroughly; feel free to revert if it causes issues.

I think we should pass the enqueuer during the instantiation phase of the various areas, rather than passing the enqueuer/logger at the last moment of responsibility like you were doing before.

I fixed the migration and added some detail to the notification.

File: coderd/workspaces.go
Comment (dannykopping) [inline coderd/workspaces.go:958]: Let's log the workspace name as well, otherwise this warning isn't very helpful.

**Review by dannykopping — APPROVED** [review]
LGTM! This looks good @BrunoQuaresma 

Have you tested with a real workspace, or just with unit tests?

**Comment by BrunoQuaresma** [comment]
> Have you tested with a real workspace, or just with unit tests?

Just unit tests, but I would like to know how to QA it in a real scenario.

**Comment by dannykopping** [comment]
> @dannykopping just to make it clear, this PR is missing tests in the lifecycle executor and templates schedule, but I'm trying to first fix the current broken tests.

Shout if you need a hand.

File: coderd/autobuild/lifecycle_executor.go
Comment (BrunoQuaresma) [reply coderd/autobuild/lifecycle_executor.go]: I saw @dannykopping use it in the notification, so I just copied it, but I think it is helpful for searching.

File: coderd/database/migrations/000227_dormancy_notification_template.up.sql
Comment (BrunoQuaresma) [reply coderd/database/migrations/000227_dormancy_notification_template.up.sql]: It was @dannykopping's refactoring, but I liked it more. It makes it easier to read, IMO.

File: coderd/database/migrations/000227_dormancy_notification_template.up.sql
Comment (BrunoQuaresma) [reply coderd/database/migrations/000227_dormancy_notification_template.up.sql]: Why?

File: coderd/database/migrations/000227_dormancy_notification_template.up.sql
Comment (dannykopping) [reply coderd/database/migrations/000227_dormancy_notification_template.up.sql]: `||` is just for concatenation; it's just for formatting in this file. The newlines are still necessary for output formatting.

File: coderd/autobuild/lifecycle_executor.go
Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor.go:335]: Log messages should generally be static, and all variability should be in fields.
I'd also recommend logging the workspace id not name since it's universally unique.

```suggestion
						e.log.Warn(e.ctx, "failed to notify of workspace marked as dormant", slog.Error(err), slog.F("workspace", dormantNotification.Workspace.Id))
```

File: coderd/schedule/mock.go
Comment (dannykopping) [inline coderd/schedule/mock.go]: Please revert this file; this was a bad refactor from my side which I fixed, but it seems you must've had a conflict and restored the changes.

File: coderd/workspaces.go
Comment (dannykopping) [inline coderd/workspaces.go:958]: Same comment as the previous case

File: coderd/schedule/mock.go
Comment (BrunoQuaresma) [reply coderd/schedule/mock.go]: This was reverted already.

File: coderd/database/migrations/000227_dormancy_notification_template.up.sql
Comment (mtojek) [reply coderd/database/migrations/000227_dormancy_notification_template.up.sql]: That's why I'm asking :) In the template you have:

`{{ if .Labels.initiator }} {{ .Labels.initiator }} {{end}}`, which suggests that _Initiator_ is optional and may not be defined. I think you can convert it to `{{ .Labels.initiator }}`. Does it sound about right, @dannykopping?

File: coderd/database/migrations/000227_dormancy_notification_template.up.sql
Comment (mtojek) [reply coderd/database/migrations/000227_dormancy_notification_template.up.sql]: I see, so it's a personal preference. Let's leave it then.

File: coderd/database/migrations/000227_dormancy_notification_template.up.sql
Comment (dannykopping) [reply coderd/database/migrations/000227_dormancy_notification_template.up.sql]: Ah, I interpreted your comment differently, and I suspect Bruno did too. It sounded like you were asking if we can drop the `initiator` field for some reason.

Yes, if the label is never blank then we don't need the conditional.

Comment (mtojek) [reply coderd/database/migrations/000227_dormancy_notification_template.up.sql]: Sorry for the shortcut thinking!

**Comment by mtojek** [comment]
@BrunoQuaresma is it still a draft PR or a regular one?

**Comment by dannykopping** [comment]
> @BrunoQuaresma is it still a draft PR or a regular one?

I made it draft so that we don't merge accidentally before testing it fully, since it has approvals already.

**Comment by BrunoQuaresma** [comment]
@dannykopping @mtojek The biggest changes since the last review:
- Added a test for the dormant notification in the lifecycle executor
- Created a different template for "Marked for Deletion" since it is a bit different from the "Marked as Dormant" use case
- Added tests to verify if notifications are sent when the dormant TTL is changed and if affected workspaces are getting notified

File: coderd/database/migrations/000230_marked_for_deletion_notification_template.up sql
Comment (dannykopping) [inline coderd/database/migrations/000230_marked_for_deletion_notification_template.up sql:14]: Nit: I'm not sure `(initiated by: autobuild)` is going to be useful to folks.
I think we can drop it from the messaging. Initiator is useful in the `workspace deleted` notification because it can be initiated by a human, but this isn't so it won't help the receiver.

File: enterprise/coderd/schedule/template.go
Comment (dannykopping) [inline enterprise/coderd/schedule/template.go:219]: Which workspace?

File: enterprise/coderd/schedule/template_test.go
Comment (dannykopping) [inline enterprise/coderd/schedule/template_test.go:629]: Ditto

File: enterprise/coderd/schedule/template_test.go
Comment (dannykopping) [inline enterprise/coderd/schedule/template_test.go:686]: Nit: explain the _why_ not the _what_; the what is right there, but the reasoning for it is what's important.

File: enterprise/coderd/schedule/template_test.go
Comment (dannykopping) [inline enterprise/coderd/schedule/template_test.go:694]: Can we be more specific here?
Which notification is being received? What are its labels? etc.
If you had this you would've caught that your new notification was never being sent.

File: coderd/database/migrations/000230_marked_for_deletion_notification_template.up sql
Comment (mtojek) [reply coderd/database/migrations/000230_marked_for_deletion_notification_template.up sql:1]: Interesting that our tests didn't catch it.

File: coderd/autobuild/lifecycle_executor.go
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor.go:338]: 👍 for removing this

File: coderd/autobuild/lifecycle_executor_test.go
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor_test.go:1114]: .. unless it is flaky, then remove this check

File: enterprise/coderd/schedule/template_test.go
Comment (mtojek) [reply enterprise/coderd/schedule/template_test.go:694]: +1

File: coderd/database/migrations/000230_marked_for_deletion_notification_template.up sql
Comment (mtojek) [reply coderd/database/migrations/000230_marked_for_deletion_notification_template.up sql:14]: +1 to drop this

**Review by mtojek — COMMENTED** [review]
I seconded a few Danny's comments.

@BrunoQuaresma If you feel a need to clarify some comments, please sync with @dannykopping. Let's not elongate this task anymore, it is already a leftover from the last sprint.

File: coderd/autobuild/lifecycle_executor.go
Comment (BrunoQuaresma) [reply coderd/autobuild/lifecycle_executor.go:338]: Idk why I put it there, thanks for noticing it.

File: coderd/autobuild/lifecycle_executor_test.go
Comment (BrunoQuaresma) [reply coderd/autobuild/lifecycle_executor_test.go:1114]: Since it is a background job, I think it is hard to find the exact time. Here is a diff between the value and the expectation:
```
expected: *time.Time(time.Date(2024, time.July, 24, 13, 37, 57, 414817000, time.UTC))
actual  : time.Time(time.Date(2024, time.July, 24, 13, 38, 56, 603858000, time.UTC))
```
I also don't think knowing if the dormant time is correct is helpful for this particular test.

File: coderd/database/migrations/000229_dormancy_notification_template.up.sql
Comment (BrunoQuaresma) [reply coderd/database/migrations/000229_dormancy_notification_template.up.sql:14]: What do you think about:
- `hours of dormancy it will be deleted` to `hours of dormancy might be deleted`
- `To prevent your workspace from being deleted, simply use it as normal.` to `To activate your workspace again, simply use it as normal.`

Comment (BrunoQuaresma) [reply coderd/database/migrations/000229_dormancy_notification_template.up.sql]: I'm moving forward with the above but we can change it later.

File: coderd/database/migrations/000229_dormancy_notification_template.up.sql
Comment (stirby) [inline coderd/database/migrations/000229_dormancy_notification_template.up.sql:14]: The wording is a little weird here. Can we change it and add a link to the dormancy docs:

I don't think we need to include the initiator, just improve the reasons (see end of message).

> Hi `{user-name}`, your workspace `{ws-name}` has been marked [dormant](https://coder.com/docs/templates/schedule#dormancy-threshold-enterprise) because of `{reason}`.

Append if dormancy cleanup is enabled and `dormancy_hours` < 24:
> Dormant workspaces are [automatically deleted](https://coder.com/docs/templates/schedule#dormancy-auto-deletion-enterprise) after `{dormancy_hours}` hours of inactivity.

Append if dormancy cleanup is enabled, and `dormancy_hours` > 24:
> Dormant workspaces are [automatically deleted](https://coder.com/docs/templates/schedule#dormancy-auto-deletion-enterprise) after `{dormancy_hours // 24}` days of inactivity.

> To prevent deletion, use your workspace with the link below.

---

The reasons also fit into differing grammatic context:
- Lifecycle executor: "breached the template's threshold for inactivity"
- API: "requested by user"
- Template: "template updated to new dormancy policy"

If we use these, they'd fit the above messages better.
- Lifcycle: "prolonged inactivity, exceeding the dormancy threshold"
- API: "a user request"
- Template: "an update to the template's dormancy"

Comment (BrunoQuaresma) [reply coderd/database/migrations/000229_dormancy_notification_template.up.sql:14]: Going to refactor the templates in the next PR to avoid further postponement.

**Pull request merged**
